Bio/Align/fasta_m8.py

```diff
@@ -0,0 +1,238 @@
+        describe a sequence alignment.
+        """
+        target_coordinates = []
+        query_coordinates = []
+        target_coordinates.append(0)
+        query_coordinates.append(0)
+        state = State.NONE
+        tokens = re.findall("([A-Z-]{2}|\\d+)", btop)
+        # each token is now
+        # - an integer
+        # - a pair of characters, which may include dashes
+        for token in tokens:
+            if token.startswith("-"):
+                if state != State.QUERY_GAP:
+                    target_coordinates.append(target_coordinates[-1])
+                    query_coordinates.append(query_coordinates[-1])
+                    state = State.QUERY_GAP
+                target_coordinates[-1] += 1
+            elif token.endswith("-"):
+                if state != State.TARGET_GAP:
+                    target_coordinates.append(target_coordinates[-1])
+                    query_coordinates.append(query_coordinates[-1])
+                    state = State.TARGET_GAP
+                query_coordinates[-1] += 1
+            else:
+                try:
+                    length = int(token)
+                except ValueError:
+                    # pair of mismatched letters
+                    length = 1
+                if state == State.MATCH:
+                    target_coordinates[-1] += length
+                    query_coordinates[-1] += length
+                else:
+                    target_coordinates.append(target_coordinates[-1] + length)
+                    query_coordinates.append(query_coordinates[-1] + length)
+                    state = State.MATCH
+        coordinates = numpy.array([target_coordinates, query_coordinates])
+        return coordinates
+
+    def parse_cigar(self, cigar):
+        """Parse a CIGAR string and return alignment coordinates.
+
+        A CIGAR string, as defined by the SAM Sequence Alignment/Map format,
+        describes a sequence alignment as a series of lengths and operation
+        (alignment/insertion/deletion) codes.
+        """
+        target_coordinates = []
+        query_coordinates = []
+        target_coordinate = 0
+        query_coordinate = 0
+        target_coordinates.append(target_coordinate)
+        query_coordinates.append(query_coordinate)
+        state = State.NONE
+        tokens = re.findall("(M|D|I|\\d+)", cigar)
+        # each token is now
+        # - the length of the operation
+        # - the operation
+        for length, operation in zip(tokens[::2], tokens[1::2]):
+            length = int(length)
+            if operation == "M":
+                target_coordinate += length
+                query_coordinate += length
+            elif operation == "I":
+                target_coordinate += length
+            elif operation == "D":
+                query_coordinate += length
+            else:
+                raise ValueError("Unexpected operation '%s'" % operation)
```

Comment:
From the narrow regex above, is this possible?

Separately, M/D/I is only a tiny subset of the SAM/BAM CIGAR language. There is also =/X as an alternative to M, and S and H for clipping. Do you anticipate adding those as needed?

Reply:
To just define the alignment, M/D/I is sufficient, and I have not seen other operations in the cigar string in output generated by FASTA. We can add other operations if somebody finds them in FASTA output.

We may want to move this code to the `Alignment` class because it is generally useful, but I want to consider that at a later stage, once we have parsers for all alignment formats we want to support.

I'd like to keep the `ValueError` for now, to help with debugging if other operations indeed appear.

Reply:
Actually, the `ValueError` cannot appear. I will remove it.

Reply:
OK - if and when this is moved for use outside of the FASTA parser, then it would need to support other CIGAR operations.